From 88c337f7910e3a9048dab1316bdb0f4a257f3213 Mon Sep 17 00:00:00 2001
From: Jernej Skrabec <jernej.skrabec@gmail.com>
Date: Sun, 12 Jul 2026 22:01:44 +0200
Subject: [PATCH] dt-bindings: display: Add Allwinner DE3 writeback

Document the real-time writeback engine found in the H6 display engine.

Signed-off-by: Jernej Skrabec <jernej.skrabec@gmail.com>
---
 .../display/allwinner,sun50i-h6-de3-wb.yaml   | 122 ++++++++++++++++++
 1 file changed, 122 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/display/allwinner,sun50i-h6-de3-wb.yaml

diff --git a/Documentation/devicetree/bindings/display/allwinner,sun50i-h6-de3-wb.yaml b/Documentation/devicetree/bindings/display/allwinner,sun50i-h6-de3-wb.yaml
new file mode 100644
index 00000000000000..9ef330bfadbc45
--- /dev/null
+++ b/Documentation/devicetree/bindings/display/allwinner,sun50i-h6-de3-wb.yaml
@@ -0,0 +1,122 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/display/allwinner,sun50i-h6-de3-wb.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Allwinner DE3 Writeback Engine
+
+maintainers:
+  - Jernej Skrabec <jernej.skrabec@gmail.com>
+
+properties:
+  compatible:
+    const: allwinner,sun50i-h6-de3-wb
+
+  reg:
+    maxItems: 1
+
+  clocks:
+    items:
+      - description: Bus clock
+      - description: Module clock
+
+  clock-names:
+    items:
+      - const: bus
+      - const: mod
+
+  iommus:
+    maxItems: 1
+
+  interrupts:
+    maxItems: 1
+
+  resets:
+    maxItems: 1
+
+  ports:
+    $ref: /schemas/graph.yaml#/properties/ports
+
+    properties:
+      port@0:
+        $ref: /schemas/graph.yaml#/$defs/port-base
+        description: Inputs from display engine mixers
+
+        properties:
+          endpoint@0:
+            $ref: /schemas/graph.yaml#/properties/endpoint
+            description: Input from mixer 0
+            properties:
+              reg:
+                const: 0
+
+          endpoint@1:
+            $ref: /schemas/graph.yaml#/properties/endpoint
+            description: Input from mixer 1
+            properties:
+              reg:
+                const: 1
+
+        anyOf:
+          - required:
+              - endpoint@0
+          - required:
+              - endpoint@1
+
+        unevaluatedProperties: false
+
+    required:
+      - port@0
+
+required:
+  - compatible
+  - reg
+  - clocks
+  - clock-names
+  - iommus
+  - interrupts
+  - resets
+  - ports
+
+additionalProperties: false
+
+examples:
+  - |
+    #include <dt-bindings/clock/sun8i-de2.h>
+    #include <dt-bindings/interrupt-controller/arm-gic.h>
+    #include <dt-bindings/reset/sun8i-de2.h>
+
+    writeback@10000 {
+        compatible = "allwinner,sun50i-h6-de3-wb";
+        reg = <0x10000 0x1000>;
+        clocks = <&display_clocks CLK_BUS_WB>,
+                 <&display_clocks CLK_WB>;
+        clock-names = "bus", "mod";
+        resets = <&display_clocks RST_WB>;
+        iommus = <&iommu 0>;
+        interrupts = <GIC_SPI 78 IRQ_TYPE_LEVEL_HIGH>;
+
+        ports {
+            #address-cells = <1>;
+            #size-cells = <0>;
+
+            port@0 {
+                #address-cells = <1>;
+                #size-cells = <0>;
+                reg = <0>;
+
+                writeback_in_mixer0: endpoint@0 {
+                    reg = <0>;
+                    remote-endpoint = <&mixer0_out_writeback>;
+                };
+
+                writeback_in_mixer1: endpoint@1 {
+                    reg = <1>;
+                    remote-endpoint = <&mixer1_out_writeback>;
+                };
+            };
+        };
+    };
+
+...
